### ISSUE TITLE
How to assert an error?

### ISSUE THREAD

**Issue by elliotchance** [vlang/v#10803]
In my tests I'm using a rather clunky and verbose way to verify an error:

```
fn is_valid() ? {
	return error("not valid")
}

fn test_is_valid() {
	mut msg := ""
	is_valid() or {
		msg = "$err"
	}
	assert msg == "not valid"
}
```

Is there a more natural way to do this?

**Comment by dumblob** [comment]
From the example I'd say you're trying to verify the **content** of an error. If that's so, why not simply:
```v
fn test_is_valid() {
	is_valid() or {
		assert "$err" == "not valid"
	}
}
```

**Comment by elliotchance** [comment]
@dumblob yes I thought about that, but that's not entirely correct because if `is_valid` does not return an error the test would pass.

It would be great if V has some native support for this, like perhaps an `err` function to capture the error as a string:

```v
// err returns the error as a string; or empty if no error occurred
// fn err(x any) string

fn test_is_valid() {
	// verify an error was returned - this only works if the empty string is treated as false
	assert err(is_valid())

	// verify an error and some check on the message
	assert err(is_valid()).contains("not_valid") 
}
```

Totally not thought out properly, just a quick brain dump.

**Comment by dumblob** [comment]
Admittedly I still kind of fail to understand what everything you want to test.

If you want to test "whether function failed", just convert the test case to panic `is_valid() or { panic( err ) }` (or maybe [in the future](https://github.com/vlang/v/issues/2030#issuecomment-778113320 ) just `is_valid() or { assert false ) }`).

If you want to check the **content** of `err`, use what I proposed above in https://github.com/vlang/v/issues/10803#issuecomment-881290664 .

If you want both, just use `assert` twice `is_valid() or { assert err == "not valid"; assert false ) }`.

If you want something else, please elaborate :wink:.

**Comment by elliotchance** [comment]
OK, let's consider a more concrete example:

```v
// create_person might add something to a database, or whatever.
pub fn create_person(name string) ? {
	if name == "" {
		return error("name cannot be empty")
	}

	// go on...
}
```

And:

```v
fn test_create_person_empty_name() {
	create_person('') or {
		assert '$err' == 'name cannot be empty'
	}
}
```

Now remove the error:

```v
pub fn create_person(name string) ? {
	// go on...
}
```

`test_create_person_empty_name` will still pass. That's the problem.

**Comment by elliotchance** [comment]
Here's a proof of concept that works but is a little clunky:

```v
fn err<T>(f T) string {
	f() or {
		return '$err'
	}

	return ''
}

fn test_create_person_empty_name() {
	assert err(fn () ? { create_person('') ? }) == 'name cannot be empty'
}
```

**Comment by dumblob** [comment]
Oh, thanks. Now I understand. I'm sorry for being so ignorant, I was led astray by all the value checking etc.

IMHO you can do that easily like this:
```v
fn test_is_valid() {
	is_valid() or { return none }
	assert false
}
```

**Comment by elliotchance** [comment]
Ah, that is better but it still has a flaw: a test would only be limited to one assertion. Or, at least the last assertion can only be an error.

**Comment by dumblob** [comment]
I don't think so :wink::
```v
fn test_is_valid() {
	is_valid() or {
		write
		any following
		tests you need
		here
		return none
	}
	assert false
}
```

**Comment by elliotchance** [comment]
Sheesh... that looks nasty. Hopefully, V can come up with something more test friendly than that.

**Comment by dumblob** [comment]
> Sheesh... that looks nasty.

I'd objectively argue it's vice versa - it makes it very explicit what it checks ("it checks the "or" branch) and what the code does (it resembles how the function will be used in practice and doesn't thus hide any bit of information nor usage).

>Hopefully, V can come up with something more test friendly than that.

Then we should discuss the type of friendliness you have in mind. For me it's still 
 bit clearer than calling `assert err( ... )` which seems to hide things (as it goes against how "conventional V is being used"). Programmers in functional languages would even tell you it's the one greatest thing you can do (as majority of functional languages have syntax leading to such programming style).

On the other hand I understand the need to use more of such "subtests" in one test (to avoid boilerplate). I think there is a chance to solve it using other existing *V* features like `defer { }` or closures (once they're supported) or generics or combination thereof. Then you can probably write some wrappers like `err()` for your own purposes.

**Comment by danieldaeschle** [comment]
```v
fn is_valid() ? {
	return error('not valid')
}

fn test_is_valid() {
	mut msg := ""
	is_valid() or {
		assert err.msg == 'not valid'
                return
	}
	assert false
}
```

**Comment by spytheman** [comment]
I've tried to abstract it out, but It is still not as clear as it could be 🤔, and the failing asserts will point to the check_error, not to the caller. I think we may need language support, like `assert` is natively supported, to clean up such patterns futher:

```v
type FnErrCB = fn () ?

fn check_error(f FnErrCB, emsg string) {
	f() or {
		assert err.msg == emsg
		return
	}
	assert false
}

fn is_valid(x int) ? {
	return error('not valid $x')
}

fn test_is_valid() {
	check_error(fn () ? { is_valid(1) ? }, 'not valid 1')
	check_error(fn () ? { is_valid(2) ? }, 'not valid 2222') 
	check_error(fn () ? { is_valid(3) ? }, 'not valid 3')
}
```

> Can this be closed then? :)
No, I think it should stay until we have a better solution.

**Comment by elliotchance** [comment]
> I'd objectively argue it's vice versa - it makes it very explicit what it checks ("it checks the "or" branch) and what the code does (it resembles how the function will be used in practice and doesn't thus hide any bit of information nor usage).
> 
> Then we should discuss the type of friendliness you have in mind. For me it's still
> bit clearer than calling `assert err( ... )` which seems to hide things (as it goes against how "conventional V is being used"). Programmers in functional languages would even tell you it's the one greatest thing you can do (as majority of functional languages have syntax leading to such programming style).

Yes, absolutely, in production code. Testing code is not production code. Unit tests purposely use incomplete values/inputs to surgically test a unit of logic, or use mocks, or some other environmental manipulation. This would be unacceptable for production code, but it's necessary for a unit test. Likewise, the tooling around unit tests is allowed to be more flexible otherwise you just end up with tons of boiler plate code in every project. Or, a dependency on third-party packages to bridge the gap.

For example, almost all Go projects I've worked with end up relying on [`testify`](https://github.com/stretchr/testify) to provide an assertion library, because the `if x != y { t.Failf("expected %v, got %v", x, y) }` is unnecessarily verbose to type for every single "assertion". In more complex (but common cases) this leads to projects having to write all their own assertions for comparing timestamps, checking if an element exists in an array, etc. So, rightfully so, people just use a mature assertion library like testify.

> On the other hand I understand the need to use more of such "subtests" in one test (to avoid boilerplate). I think there is a chance to solve it using other existing _V_ features like `defer { }` or closures (once they're supported) or generics or combination thereof. Then you can probably write some wrappers like `err()` for your own purposes.

Subtests would be nice, but still wouldn't solve this problem. It would still be pretty clunky to write a test such that several assertions need to be wrapped in individual subtests at the end to isolate each assertion.

**Comment by splace** [comment]
FYI suffering from this same problem, had already been through all above before resorting to googling and coming here.

seems the 'or' construct assumes you always branch or error, but when in testing branching on no error is a thing;  

like '!or'

**Comment by dumblob** [comment]
`! or` clashes with existing syntax (i.e. `map[x] !` and sum types).

Besides I'd find it confusing.

Btw. now you can use [inline sum types](https://github.com/vlang/v/pull/12912 ) which have the potential to address all the concerns in this thread. Feel free to give it a try.

**Comment by spytheman** [comment]
This works (you can add more constructs like that in the same test_ function), but is 5 lines per case:
```v
pub fn create_person(name string) ? {
	if name == "" {
		return error("name cannot be empty")
	}
}
fn test_create_person_empty_name() {
	if _ := create_person('') {
		assert false 
	} else {
		assert '$err' == 'name cannot be empty'
	}
}
```
